### PR TITLE
feat(eventsources/bitbucketserver): add OneEventPerChange config option for webhook event handling

### PR DESCRIPTION
**Pull request by ryancurrah**
Add OneEventPerChange config option to control whether to process each change in a repo:refs_changed webhook event as a separate event. This allows independent processing of each tag or reference update in a single webhook event, useful for triggering distinct workflows in Argo Workflows.

> Note: This is an opt-in behavior. It is not a breaking change.

<details>
  <summary>Example Webhook Event with Multiple Changes</summary>

```json
{
  "eventKey": "repo:refs_changed",
  "date": "2024-05-15T23:58:18+0000",
  "actor": {
    "name": "ryancurrah",
    "emailAddress": "ryancurrah@foobar.com",
    "id": 4127,
    "displayName": "Ryan Currah",
    "active": true,
    "slug": "ryancurrah",
    "type": "NORMAL",
    "links": {
      "self": [
        {
          "href": "https://bitbucket.foobar.com/users/ryancurrah"
        }
      ]
    }
  },
  "repository": {
    "slug": "baz",
    "id": 222,
    "name": "baz",
    "hierarchyId": "12dcc1b4996b9c76cb12",
    "scmId": "git",
    "state": "AVAILABLE",
    "statusMessage": "Available",
    "forkable": true,
    "project": {
      "key": "TEST",
      "id": 6084,
      "name": "TEST",
      "public": false,
      "type": "NORMAL",
      "links": {
        "self": [
          {
            "href": "https://bitbucket.foobar.com/projects/TEST"
          }
        ]
      }
    },
    "public": false,
    "links": {
      "clone": [
        {
          "href": "ssh://git@bitbucket.foobar.com:7999/test/baz.git",
          "name": "ssh"
        },
        {
          "href": "https://bitbucket.foobar.com/scm/test/baz.git",
          "name": "http"
        }
      ],
      "self": [
        {
          "href": "https://bitbucket.foobar.com/projects/TEST/repos/baz/browse"
        }
      ]
    }
  },
  "changes": [
    {
      "ref": {
        "id": "refs/heads/test-tag-webhooks",
        "displayId": "test-tag-webhooks",
        "type": "BRANCH"
      },
      "refId": "refs/heads/test-tag-webhooks",
      "fromHash": "0000000000000000000000000000000000000000",
      "toHash": "165bb6a516be59cfc3bde5dce04bd206a18b9e7f",
      "type": "ADD"
    },
    {
      "ref": {
        "id": "refs/tags/v0.0.0-1",
        "displayId": "v0.0.0-1",
        "type": "TAG"
      },
      "refId": "refs/tags/v0.0.0-1",
      "fromHash": "0000000000000000000000000000000000000000",
      "toHash": "8b473721c4cad133e9cb3917c7b705d541041cdc",
      "type": "ADD"
    }
  ]
}
```
</details>

The above ☝️ webhook event would be turned into two `BitbucketServerEventData` events, `.changes[]` will not have a length greater than one.

<details>
  <summary>Split Up Webhook Event 1 (Shortened for brevity)</summary>

```json
{
  "eventKey": "repo:refs_changed",
  "date": "2024-05-15T23:58:18+0000",
  "actor": {
    "name": "ryancurrah",
    "emailAddress": "ryancurrah@foobar.com",
    "id": 4127,
    "displayName": "Ryan Currah",
..........................................
    }
  },
  "changes": [
    {
      "ref": {
        "id": "refs/heads/test-tag-webhooks",
        "displayId": "test-tag-webhooks",
        "type": "BRANCH"
      },
      "refId": "refs/heads/test-tag-webhooks",
      "fromHash": "0000000000000000000000000000000000000000",
      "toHash": "165bb6a516be59cfc3bde5dce04bd206a18b9e7f",
      "type": "ADD"
    }
  ]
}
```
</details>

<details>
  <summary>Split Up Webhook Event 2 (Shortened for brevity)</summary>

```json
{
  "eventKey": "repo:refs_changed",
  "date": "2024-05-15T23:58:18+0000",
  "actor": {
    "name": "ryancurrah",
    "emailAddress": "ryancurrah@foobar.com",
    "id": 4127,
    "displayName": "Ryan Currah",
..........................................
    }
  },
  "changes": [
    {
      "ref": {
        "id": "refs/tags/v0.0.0-1",
        "displayId": "v0.0.0-1",
        "type": "TAG"
      },
      "refId": "refs/tags/v0.0.0-1",
      "fromHash": "0000000000000000000000000000000000000000",
      "toHash": "8b473721c4cad133e9cb3917c7b705d541041cdc",
      "type": "ADD"
    }
  ]
}
```
</details>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-events/blob/master/USERS.md).

### CODE REVIEW

File: pkg/apis/eventsource/v1alpha1/types.go [hunk 1]
```diff
@@ -1051,7 +1051,7 @@ type BitbucketAuth struct {
 	OAuthToken *corev1.SecretKeySelector `json:"oauthToken,omitempty" protobuf:"bytes,2,opt,name=oauthToken"`
 }
 
-// BasicAuth holds the information required to authenticate user via basic auth mechanism
+// BitbucketBasicAuth holds the information required to authenticate user via basic auth mechanism
```
Comment (whynowy) [inline pkg/apis/eventsource/v1alpha1/types.go:1054]: Thanks!

File: eventsources/sources/bitbucketserver/start.go [hunk 2]
```diff
@@ -709,3 +670,165 @@ func (c *customBitbucketClient) GetCommitPullRequests(project, repository, commi
...
+//
+// For valid "repo:refs_changed" events, it:
+//   - Unmarshals the JSON body into a refsChangedWebhookEvent struct.
+//   - Optionally filters out changes associated with open pull requests if SkipBranchRefsChangedOnOpenPR is enabled.
+//   - Depending on the OneEventPerChange setting, it either batches all changes into one event or creates individual events for each change.
+//
+// It returns a slice of byte slices containing the processed or original webhook events, and any error encountered during processing.
+func webhookHandler(request *http.Request, logger *zap.SugaredLogger, body []byte, bitbucketserverEventSource *v1alpha1.BitbucketServerEventSource, router *Router) ([][]byte, error) {
+	if request.Header.Get("X-Event-Key") != "repo:refs_changed" {
+		// Don't continue if this is not a repo:refs_changed webhook event.
```
Comment (whynowy) [inline eventsources/sources/bitbucketserver/start.go:686]: Is this the same as the previous logic?

Comment (ryancurrah) [reply eventsources/sources/bitbucketserver/start.go:686]: Pretty much the same except I learned that Bitbucket sends the event type/key in a HTTP header named `X-Event-Key`. Which means we don't have to decode every event now. Making the code easier to read and little more performant.

Screenshot 2024-05-29 at 10 21 33 PM
Screenshot 2024-05-29 at 10 23 23 PM

Comment (ryancurrah) [reply eventsources/sources/bitbucketserver/start.go:686]: @whynowy does that explanation help?

**Comment by ryancurrah** [comment]
Ive been using this change with and without webhooks in production and it is working fine.

**Comment by ryancurrah** [comment]
Hey @whynowy just a note we have been running this change in production for a while now with no issues. Any chance you can review it?

File: eventsources/sources/bitbucketserver/custombitbucketserverclient.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+func (p *pagination) StartStr() string {
+	return strconv.Itoa(p.Start)
+}
+
+func (p *pagination) LimitStr() string {
+	return strconv.Itoa(p.Limit)
+}
+
+func (c *customBitbucketServerClient) authHeader(req *http.Request) {
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.token))
```
Comment (dekaiser123) [inline eventsources/sources/bitbucketserver/custombitbucketserverclient.go:52]: Some BBservers are only configured for Basic authentication even with the use of PA tokens. Is it possible change the Header set based on the flag under the accessToken section in https://github.com/argoproj/argo-events/blob/master/examples/event-sources/bitbucketserver.yaml#L37

**Review by dekaiser123 — CHANGES_REQUESTED** [review]

**Comment by ryancurrah** [comment]
@whynowy ???

File: eventsources/sources/bitbucketserver/custombitbucketserverclient.go (hunk 3 above)
Comment (ryancurrah) [reply eventsources/sources/bitbucketserver/custombitbucketserverclient.go:52]: This change is big enough we should open a new PR for that.

File: pkg/apis/events/v1alpha1/eventsource_types.go [hunk 4]
```diff
@@ -1044,7 +1052,10 @@ type BitbucketServerRepository struct {
 }
 
 func (b BitbucketServerEventSource) ShouldCreateWebhooks() bool {
-	return b.AccessToken != nil && b.Webhook != nil && b.Webhook.URL != ""
+	return b.AccessToken != nil &&
+		b.Webhook != nil &&
+		b.Webhook.URL != "" &&
+		((len(b.GetBitbucketServerRepositories()) > 0 && b.GetBitbucketServerRepositories() != nil) || len(b.Projects) > 0)
```
Comment (whynowy) [inline pkg/apis/events/v1alpha1/eventsource_types.go:1058]: Nit: `len(b.GetBitbucketServerRepositories()) > 0` covers `b.GetBitbucketServerRepositories() != nil`.

File: pkg/apis/eventsource/v1alpha1/openapi_generated.go [hunk 5]
```diff
@@ -0,0 +1,4642 @@
+//go:build !ignore_autogenerated
```
Comment (whynowy) [inline pkg/apis/eventsource/v1alpha1/openapi_generated.go:1]: this file should not be here.

Comment (ryancurrah) [reply pkg/apis/eventsource/v1alpha1/openapi_generated.go:1]: @whynowy ack. The rebase was a little difficult. Will delete this file.

**Review by whynowy — APPROVED** [review]
Thanks! I assume it has been well tested.

**Comment by ryancurrah** [comment]
Yeah we have been using this branch for a while now. It will also resolve an issue @tooptoop4 reported.

**Comment by tooptoop4** [comment]
when is next release? 🤞

**Comment by tooptoop4** [comment]
@ryancurrah 
there seem to be 2 issues when using just skipBranchRefsChangedOnOpenPR: true config ->

1. if changing refs on a closed pr then reopening it, get 2 wfs started, one with "eventKey":"pr:from_ref_updated" and one with "eventKey":"pr:opened"....expecting just 1
2. if changing refs on an already open pr am getting no wfs started, ....expecting to get 1 (i guess the config is as named but there seems to be no way to achieve getting just wf whenever code has changed or very first time creating PR)

**Comment by ryancurrah** [comment]
1. I am able to reproduce that scenario, it's a specific set of steps that need to happen to trigger it. I do not think it is a major issue considering it's not too often someone is going to push a commit and re-open a closed PR. We also have workflow avoidance configured so a second workflow running against the same commit will cancel the first one.
2. You need separate webhooks for branches/tags and pull requests.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: EventSource
metadata:
  name: bitbucketserver
spec:
  service:
    ports:
      - port: 12000
        targetPort: 12000
  bitbucketserver:
    commit-pushed:
      oneEventPerChange: true
      skipBranchRefsChangedOnOpenPR: true
      webhook:
        endpoint: /commit-pushed
    pr-created-updated:
      webhook:
        endpoint: /pr-created-updated
```

**Comment by tooptoop4** [comment]
@ryancurrah can u share the yaml to achieve "We also have workflow avoidance configured so a second workflow running against the same commit will cancel the first one." ?

**Comment by ryancurrah** [comment]
It's all in Go code. It's something we built ourself with some advice from Pipekit.

**Pull request merged**